[SelectionDAG] Move VSelect sign pattern check from AArch64 to general SelectionDAG

LumioseSil opened the pull request:
For some reason the check is already there, but it bails out. Doing the transform in SelDAG has no negative effect.

topperc commented:
> For some reason the check is already there, but it bails out. Probably to allow aarch64 to transform it, but why not have the logic in SelectionDAG then?

Have you consulted git blame and git log to see if you can find an answer?

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -13143,7 +13143,13 @@ static SDValue combineVSelectWithAllOnesOrZeros(SDValue Cond, SDValue TVal,
+    SDValue LHS = Cond.getOperand(0);
+    SDValue ShiftC = DAG.getShiftAmountConstant(VT.getScalarSizeInBits() - 1,
+                                                VT.getScalarType(), DL);
+    SDValue ShiftAmount = DAG.getSplatBuildVector(VT, DL, ShiftC);
```
getShiftAmountConstant should be able to splat for you:
```
SDValue ShiftAmount =DAG.getShiftAmountConstant(VT.getScalarSizeInBits() - 1, VT, DL);
```

AZero13 commented:
> > For some reason the check is already there, but it bails out. Probably to allow aarch64 to transform it, but why not have the logic in SelectionDAG then?
> 
> Have you consulted git blame and git log to see if you can find an answer?

I did actually.

LumioseSil committed:
Pre-commit test (NFC)

AZero13 commented:
Do I have to check that the operation is legal too or no?

topperc commented:
> > > For some reason the check is already there, but it bails out. Probably to allow aarch64 to transform it, but why not have the logic in SelectionDAG then?
> > 
> > 
> > Have you consulted git blame and git log to see if you can find an answer?
> 
> I did actually.

And that didn't give any answers?

AZero13 commented:
> > > > For some reason the check is already there, but it bails out. Probably to allow aarch64 to transform it, but why not have the logic in SelectionDAG then?
> > > 
> > > 
> > > Have you consulted git blame and git log to see if you can find an answer?
> > 
> > 
> > I did actually.
> 
> And that didn't give any answers?

It was added last week, so no it did not.

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
Use getShiftAmountConstant

RKSimon reviewed (changes requested):
Address merge failures

RKSimon commented on llvm/test/CodeGen/X86/cmp-select-sign.ll:
Add nounwind (preferably to all tests for consistency) to silence cfi noise
```suggestion
define <4 x i32> @sign_4xi32_multi_use(<4 x i32> %a) nounwind {
```

LumioseSil committed:
[SelectionDAG] Move VSelect sign pattern check from AArch64 general SelectionDAG

For some reason the check is already there, but it bails out. Doing the transform in SelDAG has no negative effect.

AZero13 commented:
Done @RKSimon

arsenm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -13957,8 +13957,13 @@ static SDValue combineVSelectWithAllOnesOrZeros(SDValue Cond, SDValue TVal,
-      ISD::isConstantSplatVectorAllOnes(FVal.getNode())) {
-    return SDValue();
+      ISD::isConstantSplatVectorAllOnes(FVal.getNode()) &&
+      !TLI.shouldAvoidTransformToShift(VT, VT.getScalarSizeInBits() - 1)) {
```
This is one of those terrible TLI hooks that don't really make any sense, but I see it's preexisting

RKSimon commented on llvm/test/CodeGen/X86/cmp-select-sign.ll:
```diff
@@ -0,0 +1,1003 @@
+  ret i64 %res
+}
+
+define i64 @not_sign_i64_2(i64 %a) {
```
(style) don't number tests - use descriptive names (e.g. not_sign_i64_sgt)

same for the others

AZero13 commented:
Done @RKSimon

RKSimon reviewed (approved):
LGTM

AZero13 commented:
> LGTM

Thank you can you please merge?

AZero13 commented:
@RKSimon

RKSimon committed:
Merge branch 'main' into thumb2
